EIPS/eip-3436.md

```diff
@@ -0,0 +1,119 @@
+---
+eip: 3436
+title: Expanded Clique Block Choice Rule
+author: Danno Ferrin (@shemnon)
+discussions-to: https://ethereum-magicians.org/t/eip-3436-expanded-clique-block-choice-rule/5809
+status: Draft
+type: Standards Track
+category: Networking
+created: 2021-03-25
+requires: 225
+---
+
+## Simple Summary
+
+Add a four step block rule to Clique that should reduce block production deadlocks
+
+## Abstract
+
+The current specification of Clique allows for multiple competing blocks from producers but does not
+provide any strategies to pick blocks aside from the current "highest total difficulty" rule. This
+EIP proposes a four step choice rule of highest total difficulty, shortest chain, most recently
+in-turn, and lowest hash. This would prevent deadlocks that have occurred in production systems.
+
+## Motivation
+
+Recently (5 March 2021) there were a couple of deadlocks in the Goerli multi-client Clique network.
+The number of active validators was greater than 1/2 of the available validators so a chain halt
+should not have occurred. The halt was resolved by an inactive validator coming back on line. The
+state of the chain was in one of two configurations of 8 validators that can result in a chain halt.
+Three of the four clients observed a choice sequence of lowest total difficulty followed by first
+observed block. Geth added one extra rule of preferring the shortest chain before preferring the
+first observed block. This fork would have resolved itself with Geth's rule, but there is still a
+configuration where the chain can halt with a shortest chain rule.
+
+## Specification
+
+When a Click validator is choosing between two different chain head blocks to build a nre proposed
+block they should chose the blocks from the following priority list.
+
+1. Choose the block with the most total difficulty.
+2. Then choose the block with the lowest block number.
+3. Then choose the block that has had the least recent in-turn slot.
+4. Then choose the block with the lowest hash.
+
+When resolving rule 3 clients should use the following formula, where `Vi` is the integer index of
+the validator that signed the block when sorted as per epoch checkpointing, `Hn` is the number of
+the header, and `Vt` is the count of the current validators. Clients should choose the block with
+the **smallest** value.
+
+```
+(Hn - Vi) % Vt
```

Comment:
Recommend human readable variable names rather than single or double letter names:
```suggestion
(header_number - validator_index) % validator_count
```